quinn-h3/src/frame.rs

```diff
@@ -158,12 +158,14 @@ impl Future for WriteFrame {
                 WriteFrameState::Payload(ref mut send, ref mut p) => {
                     let wrote = ready!(Pin::new(send).poll_write(cx, p))?;
                     p.advance(wrote);
-                    let send = match mem::replace(&mut self.state, WriteFrameState::Finished) {
-                        WriteFrameState::Payload(s, _) => s,
-                        _ => unreachable!(),
-                    };
-                    self.state = WriteFrameState::Finished;
-                    return Poll::Ready(Ok(send));
+                    if p.is_empty() {
```

Comment:
Both here and in the preceding block, it might help readability to invert the `is_empty()` condition and do an early return if not empty (this can be a follow-up commit).

Reply:
Also, how hard would it be to add a test that covers this?

Reply:
It is doable I guess, making a trait out of the used methods of `SendStream` in `WriteFrame`, then using `WriteFrame<T: SendStreamTrait>`. So the tests can pass a write-to-buffer implementation. Some parameters for this mocked impl could then be used to define some `SendStream` desired testing behavior. I'm not sure how hard can this be.

There might be simpler solutions?

Testing all edge cases will be a must-have in the future of this crate .

Reply:
Do you want those tests included with this PR?

Reply:
Nah, but maybe write an issue about it so we don't forget if you don't intend to do it nowish?